File: packages/sdks/server-sdk/src/services/PopupNoticeService.ts
/**
 * Popup Notice Service
 * Handles in-game popup notice retrieval
 * Uses per-environment API pattern: GET /api/v1/server/:env/ingame-popup-notices
 * Extends BaseEnvironmentService for common fetch/caching logic
 */

import { ApiClient } from '../client/ApiClient';
import { Logger } from '../utils/logger';
import { PopupNotice } from '../types/api';
import { BaseEnvironmentService } from './BaseEnvironmentService';

// Response type is directly PopupNotice[] (not wrapped in object)
type PopupNoticeListResponse = PopupNotice[];

export class PopupNoticeService extends BaseEnvironmentService<PopupNotice, PopupNoticeListResponse, number> {
  constructor(apiClient: ApiClient, logger: Logger, defaultEnvironment: string = 'development') {
    super(apiClient, logger, defaultEnvironment);
  }

  // ==================== Abstract Method Implementations ====================

  protected getEndpoint(environment: string): string {
    return `/api/v1/server/${encodeURIComponent(environment)}/ingame-popup-notices`;
  }

  protected extractItems(response: PopupNoticeListResponse): PopupNotice[] {
    return response; // Response is directly an array
  }

  protected getServiceName(): string {
    return 'popup notices';
  }

  protected getItemId(item: PopupNotice): number {
    return item.id;
  }

  // ==================== Domain-specific Methods ====================

  /**
   * Get popup notice by ID
   * GET /api/v1/server/:env/ingame-popup-notices/:id
   */
  async getById(id: number, environment?: string): Promise<PopupNotice> {
    const env = environment || this.defaultEnvironment;
    this.logger.debug('Fetching popup notice by ID', { id, environment: env });

    const response = await this.apiClient.get<{ notice: PopupNotice }>(`/api/v1/server/${encodeURIComponent(env)}/ingame-popup-notices/${id}`);

    if (!response.success || !response.data) {
      throw new Error(response.error?.message || 'Failed to fetch popup notice');
    }

    this.logger.info('Popup notice fetched', { id });

    return response.data.notice;
  }

  /**
   * Update a single popup notice in cache (immutable)
   * If isVisible is false, removes the notice from cache (no API call needed)
   * If isVisible is true but not in cache, fetches and adds it to cache
   * If isVisible is true and in cache, fetches and updates it
   */
  async updateSingleNotice(id: number, environment?: string, isVisible?: boolean | number): Promise<void> {
    try {
      this.logger.debug('Updating single popup notice in cache', { id, environment, isVisible });

      const envKey = environment || this.defaultEnvironment;

      // If isVisible is explicitly false (0 or false), just remove from cache
      if (isVisible === false || isVisible === 0) {
        this.logger.info('Popup notice isVisible=false, removing from cache', { id, environment: envKey });
        this.removeFromCache(id, environment);
        return;
      }

      // Otherwise, fetch from API and add/update
      // Add small delay to ensure backend transaction is committed
      await new Promise(resolve => setTimeout(resolve, 100));

      // Fetch the single notice from backend using getById
      let updatedNotice: PopupNotice;
      try {
        updatedNotice = await this.getById(id, environment);
      } catch (_error: any) {
        // If notice not found (404), it's no longer active or visible
        this.logger.debug('Popup notice not found or not active, removing from cache', { id, environment: envKey });
        this.removeFromCache(id, environment);
        return;
      }

      this.updateItemInCache(updatedNotice, environment);
    } catch (error: any) {
      this.logger.error('Failed to update single popup notice in cache', {
        id,
        environment,
        error: error.message,
      });
      // If update fails, fall back to full refresh
      await this.refresh();
    }
  }

  /**
   * Remove a popup notice from cache (immutable)
   * @deprecated Use removeFromCache instead
   */
  removeNotice(id: number, environment?: string): void {
    this.removeFromCache(id, environment);
  }

  /**
   * Get active notices for a specific world
   */
  getNoticesForWorld(worldId: string, environment?: string): PopupNotice[] {
    const notices = this.getCached(environment);
    return notices.filter((notice) => {
      if (!notice.targetWorlds || notice.targetWorlds.length === 0) {
        return true; // No targeting = show to all
      }
      return notice.targetWorlds.includes(worldId);
    });
  }

  /**
   * Get active popup notices that are currently visible for the given context
   * Filters by date range (startDate/endDate) and targeting fields
   * Sorted by displayPriority (ascending, lower values mean higher priority)
   * @returns Array of active popup notices, empty array if none match
   */
  getActivePopupNotices(options?: {
    platform?: string;
    channel?: string;
    subChannel?: string;
    worldId?: string;
    userId?: string;
    environment?: string;
  }): PopupNotice[] {
    const now = new Date();
    const { platform, channel, subChannel, worldId, userId, environment } = options ?? {};

    const notices = this.getCached(environment);
    const filtered = notices.filter((notice) => {
      // Check startDate: if set, current time must be after startDate
      if (notice.startDate) {
        const startDate = new Date(notice.startDate);
        if (now < startDate) {
          return false;
        }
      }

      // Check endDate: if set, current time must be before endDate
      if (notice.endDate) {
        const endDate = new Date(notice.endDate);
        if (now > endDate) {
          return false;
        }
      }

      // Platform targeting
      if (platform && notice.targetPlatforms && notice.targetPlatforms.length > 0) {
        const isInPlatformList = notice.targetPlatforms.includes(platform);
        const inverted = Boolean(notice.targetPlatformsInverted);
        if (inverted ? isInPlatformList : !isInPlatformList) {
          return false;
        }
      }

      // Channel targeting
      if (channel && notice.targetChannels && notice.targetChannels.length > 0) {
        const isInChannelList = notice.targetChannels.includes(channel);
        const inverted = Boolean(notice.targetChannelsInverted);
        if (inverted ? isInChannelList : !isInChannelList) {
          return false;
        }
      }

      // Subchannel targeting (format: channel:subchannel)
      if (
        channel &&
        subChannel &&
        notice.targetSubchannels &&
        notice.targetSubchannels.length > 0
      ) {
        const subchannelKey = `${channel}:${subChannel}`;
        const isInSubchannelList = notice.targetSubchannels.includes(subchannelKey);
        const inverted = Boolean(notice.targetSubchannelsInverted);
        if (inverted ? isInSubchannelList : !isInSubchannelList) {
          return false;
        }
      }

      // World targeting
      if (worldId && notice.targetWorlds && notice.targetWorlds.length > 0) {
        const isInWorldList = notice.targetWorlds.includes(worldId);
        const inverted = Boolean(notice.targetWorldsInverted);
        if (inverted ? isInWorldList : !isInWorldList) {
          return false;
        }
      }

      // User targeting
      if (userId && notice.targetUserIds) {
        const userIdList = Array.isArray(notice.targetUserIds)
          ? notice.targetUserIds
          : String(notice.targetUserIds)
              .split(',')
              .map((id) => id.trim())
              .filter((id) => id);

        if (userIdList.length > 0) {
          const isInUserList = userIdList.includes(userId);
          const inverted = Boolean(notice.targetUserIdsInverted);
          if (inverted ? isInUserList : !isInUserList) {
            return false;
          }
        }
      }

      return true;
    });

    return filtered.sort((a, b) => a.displayPriority - b.displayPriority);
  }
}

File: packages/sdks/server-sdks/gatrix-dotnet-server-sdk/src/Gatrix.Server.Sdk/Services/PopupNoticeService.cs
using System.Text.Json.Serialization;
using Gatrix.Server.Sdk.Cache;
using Gatrix.Server.Sdk.Client;
using Gatrix.Server.Sdk.Models;
using Microsoft.Extensions.Logging;

namespace Gatrix.Server.Sdk.Services;

public interface IPopupNoticeService
{
    Task InitializeAsync(string environmentId, CancellationToken ct = default);
    Task<List<PopupNotice>> FetchAsync(string environmentId, CancellationToken ct = default);
    List<PopupNotice> GetCached(string environmentId);
    List<PopupNotice> GetAll(string environmentId);
    List<PopupNotice> GetForWorld(string worldId, string environmentId);
    List<PopupNotice> GetActive(string environmentId, string? platform = null,
        string? channel = null, string? worldId = null, string? userId = null);
    Task UpdateSingleNoticeAsync(int id, string environmentId, bool? isVisible = null, CancellationToken ct = default);
    void RemoveNotice(int id, string environmentId);
}

internal class PopupNoticeByIdResponse
{
    [JsonPropertyName("notice")] public PopupNotice Notice { get; set; } = null!;
}

public class PopupNoticeService : BaseEnvironmentService<PopupNotice, List<PopupNotice>>, IPopupNoticeService
{
    public PopupNoticeService(GatrixApiClient apiClient, ILogger<PopupNoticeService> logger, ICacheStorageProvider? storage = null)
        : base(apiClient, logger, storage) { }

    protected override string ServiceName => "PopupNotice";
    protected override string GetEndpoint(string environmentId) =>
        $"/api/v1/server/ingame-popup-notices";
    protected override List<PopupNotice> ExtractItems(List<PopupNotice> response) => response;
    protected override object GetItemId(PopupNotice item) => item.Id;

    public Task<List<PopupNotice>> FetchAsync(string environmentId, CancellationToken ct = default) =>
        FetchByEnvironmentAsync(environmentId, ct);
    public List<PopupNotice> GetAll(string environmentId) => GetCached(environmentId);

    // ── Single-item cache operations (event-driven) ─────────────

    public async Task UpdateSingleNoticeAsync(int id, string environmentId, bool? isVisible = null, CancellationToken ct = default)
    {
        try
        {
            if (isVisible == false)
            {
                Logger.LogInformation("PopupNotice isVisible=false, removing from cache (id={Id})", id);
                RemoveFromCache(id, environmentId);
                return;
            }

            await Task.Delay(100, ct);

            var response = await ApiClient.GetAsync<PopupNoticeByIdResponse>(
                $"/api/v1/server/ingame-popup-notices/{id}", ct: ct);

            if (!response.Success || response.Data?.Notice is null)
            {
                Logger.LogDebug("PopupNotice not found (id={Id}), removing from cache", id);
                RemoveFromCache(id, environmentId);
                return;
            }

            UpsertItemInCache(response.Data.Notice, environmentId);
            Logger.LogDebug("Single PopupNotice upserted in cache (id={Id})", id);
        }
        catch (Exception ex)
        {
            Logger.LogError(ex, "Failed to update single PopupNotice (id={Id}), falling back to full refresh", id);
            await FetchByEnvironmentAsync(environmentId, ct);
        }
    }

    public void RemoveNotice(int id, string environmentId)
    {
        RemoveFromCache(id, environmentId);
        Logger.LogInformation("PopupNotice removed from cache (id={Id})", id);
    }

    // ── Targeting ───────────────────────────────────────────────

    public List<PopupNotice> GetForWorld(string worldId, string environmentId) =>
        GetCached(environmentId).Where(n => MatchesTarget(n.TargetWorlds, n.TargetWorldsInverted, worldId)).ToList();

    public List<PopupNotice> GetActive(string environmentId, string? platform = null,
        string? channel = null, string? worldId = null, string? userId = null)
    {
        var now = DateTime.UtcNow;
        return GetCached(environmentId)
            .Where(n =>
            {
                if (n.StartDate is not null && DateTime.Parse(n.StartDate) > now) return false;
                if (n.EndDate is not null && DateTime.Parse(n.EndDate) < now) return false;
                if (platform is not null && !MatchesTarget(n.TargetPlatforms, n.TargetPlatformsInverted, platform)) return false;
                if (channel is not null && !MatchesTarget(n.TargetChannels, n.TargetChannelsInverted, channel)) return false;
                if (worldId is not null && !MatchesTarget(n.TargetWorlds, n.TargetWorldsInverted, worldId)) return false;
                if (userId is not null && !MatchesTarget(n.TargetUserIds, n.TargetUserIdsInverted, userId)) return false;
                return true;
            })
            .OrderBy(n => n.DisplayPriority)
            .ToList();
    }

    private static bool MatchesTarget(List<string>? targets, bool? inverted, string value)
    {
        if (targets is null or { Count: 0 }) return true;
        var contains = targets.Contains(value, StringComparer.OrdinalIgnoreCase);
        return inverted == true ? !contains : contains;
    }
}
